### PR TITLE
#5011 Fix: Image with CID (Content-ID) not displaying in pgp block

### LINKED ISSUE #5011: Image with CID (Content-ID) not displaying in pgp block
Here's image src format

\ 

image

### PR DESCRIPTION
**Pull request by ioanmo226**
This PR fixed image with CID (Content-ID) not displaying in pgp block issue

close #5011 // if this PR closes an issue

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

File: extension/js/common/platform/xss.ts [hunk 1]
```diff
@@ -131,7 +131,7 @@ export class Xss {
           img.remove(); // just skip images
         } else if (!src) {
           img.remove(); // src that exists but is null is suspicious
-        } else if (imgHandling === 'IMG-TO-LINK') {
+        } else if (imgHandling === 'IMG-TO-LINK' && !img.classList.contains('cid_to_base64_img')) {
```
Comment (sosnovsky) [inline extension/js/common/platform/xss.ts:134]: I think we should have consistent behavior for inline images. As currently images with cid will be shown immediately on load, while other inline images will need to be clicked before showing (with `IMG-TO-LINK` image handling)

**Review by sosnovsky — COMMENTED** [review]
Implementation looks good 👍

Need also test how this cid images are converted when creating reply or forwarding message (they should be replaced just with their names)

File: extension/js/common/platform/xss.ts (hunk 1 above)
Comment (ioanmo226) [reply extension/js/common/platform/xss.ts:134]: I was also considering that. Should we display the 'show inline' button and then show the image when the user clicks on it? Or should we simply show the inline image immediately? I don't think there are any security issues with showing the inline image right away, though.

File: extension/chrome/elements/pgp_block_modules/pgp-block-render-module.ts [hunk 2]
```diff
@@ -319,6 +321,50 @@ export class PgpBlockViewRenderModule {
...
+
+      // Check if the src attribute contains a CID
+      if (matches && matches[1]) {
+        const contentId = matches[1];
+        const contentIdAttachment = attachments.find(attachment => attachment.cid === `<${contentId}>`);
+
+        // Replace the src attribute with a base64 encoded string
+        if (contentIdAttachment) {
+          inlineCIDAttachments.push(contentIdAttachment);
+          node.setAttribute('src', `data:image/png;base64,${contentIdAttachment.getData().toBase64Str()}`);
```
[the remark this answers is not in the document]
Comment (ioanmo226) [reply extension/chrome/elements/pgp_block_modules/pgp-block-render-module.ts:352]: Sounds good 👍

File: extension/js/common/platform/xss.ts (hunk 1 above)
Comment (sosnovsky) [reply extension/js/common/platform/xss.ts:134]: Yes, I think showing inline images immediately without additional button will be more convenient

**Comment by martgil** [comment]
> I need a hand creating a test email for the UI testing, specifically one that includes an "img src" with a content ID.

@ioanmo226 sir, here are the email for test with CID images.

https://mail.google.com/mail/u/flowcrypt.compatibility@gmail.com/#inbox/KtbxLrjCNLMzfJBPGVJkSrFKTCGQfvFRkL - unencrypted email.
https://mail.google.com/mail/u/flowcrypt.compatibility@gmail.com/#inbox/FMfcgzGslbKpLhNpDphbRqtNRQgxkXNv - encrypted mime message.

File: extension/js/common/platform/xss.ts (hunk 1 above)
Comment (ioanmo226) [reply extension/js/common/platform/xss.ts:134]: But then do we need `IMG-TO-LINK` type?
Do we have to remove this logic and update UI test too?

https://github.com/FlowCrypt/flowcrypt-browser/blob/6ff785cc55d3f892ffdea8609871628f8cfb2884/extension/js/common/platform/xss.ts#L9

Comment (sosnovsky) [reply extension/js/common/platform/xss.ts:134]: Seems like `IMG-TO-LINK` isn't needed after this change.

And we're already showing remote images immediately after load, so this update will make all images to behave the same - inline and remote.

Comment (ioanmo226) [reply extension/js/common/platform/xss.ts:134]: Yeah, I also agree

Comment (ioanmo226) [reply extension/js/common/platform/xss.ts:134]: But one thing I'm unsure is that `show image` button was customer request before then.

Comment (ioanmo226) [reply extension/js/common/platform/xss.ts:134]: Maybe @tomholub, I think you can confirm why we displayed `show image` button before.

Comment (sosnovsky) [reply extension/js/common/platform/xss.ts:134]: Yeah, I was thinking about it, but probably it was for security reasons (as for remote images).
But as we allow to load remote images now, then it won't be an issue to have inline images to be auto-showed too

Comment (ioanmo226) [reply extension/js/common/platform/xss.ts:134]: Makes sense

**Comment by ioanmo226** [comment]
@sosnovsky Could you explain more?
For example, what should be reply/forward content for this sample email.
> Need also test how this cid images are converted when creating reply or forwarding message (they should be replaced just with their names)

https://mail.google.com/mail/u/0/#inbox/FMfcgzGslbKpLlkkQNMnMhHszzTxCRpk

**Comment by ioanmo226** [comment]
You mean image name or cid link?

**Comment by ioanmo226** [comment]
We show remote image url when replying/forwarding though for remote images.

**Comment by sosnovsky** [comment]
It should look like other inline images - just image name (without url)

Screenshot 2023-03-17 at 20 52 50

**Comment by ioanmo226** [comment]
Got it

**Comment by ioanmo226** [comment]
@sosnovsky Ready for reveiw!

**Review by sosnovsky — APPROVED** [review]
Works well now 👍

**Pull request merged**